blockchain/validators/src/metadata_key.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2018-20&&22 The MobileCoin Foundation
+
+//! Block metadata key validator.
+
+use crate::{Error, NodeKeyValidity};
+use mc_blockchain_types::{BlockIndex, BlockMetadata};
+use mc_crypto_keys::Ed25519Public;
+use std::collections::{BTreeMap, BTreeSet};
+
+/// A validator for block metadata signing keys and signatures.
+#[derive(Clone, Debug)]
+pub struct MetadataKeyValidator {
+    keys: BTreeMap<Ed25519Public, BTreeSet<NodeKeyValidity>>,
+}
+
+impl MetadataKeyValidator {
+    /// Instantiate a validator with the given node key validity ranges.
+    pub fn new(configs: &[NodeKeyValidity]) -> Self {
+        let mut keys = BTreeMap::<Ed25519Public, BTreeSet<NodeKeyValidity>>::new();
+        for config in configs {
+            keys.entry(config.message_signing_pub_key)
+                .or_default()
+                .insert(config.clone());
+        }
+
+        Self { keys }
+    }
+
+    /// Check whether the given block metadata is valid at the given index.
+    pub fn verify(&self, metadata: &BlockMetadata, block_index: BlockIndex) -> Result<(), Error> {
```

Comment:
I don't think I have the full picture and maybe it depends on how the key -> ranges are built up and what trust needs to be there.

Let us say that Key A is valid from blocks 20-30, so this function gets in index 25 with key A.  This method would return true.  But what if A was only used for block 20 and block 30, but nothing inbetween?

It could be that "verify" and "valid" might need to be explained more for this method.  I.e. is it that this is a potential _candidate_ and more looking needs to be done to ensure it really is correct for this block

Reply:
Fair point. I elaborated in the method docs.

FWIW the intent here is that `LedgerDB` will call this as part of its validation before appending blocks, and this config can be used to revoke signing keys.